File: packages/schema_runtime_ts/src/schema/ref_resolver.ts
import {
  type ComponentNode,
  type RefNode,
  type SchemaNode,
  type ScreenSchema,
  makeComponentNode,
} from './schema_models.js';
import { parseFragmentSchema } from './schema_parser.js';

export interface FragmentDocumentLoader {
  loadFragmentDocument(fragmentId: string): Promise<Record<string, unknown>>;
}

export interface RefResolutionError {
  readonly path: string;
  readonly ref: string;
  readonly message: string;
}

export interface RefResolutionResult {
  readonly schema: ScreenSchema;
  readonly errors: RefResolutionError[];
  readonly isFullyResolved: boolean;
}

function isRefNode(node: SchemaNode): node is RefNode {
  return (node as RefNode).kind === 'ref';
}

function isComponentNode(node: SchemaNode): node is ComponentNode {
  return (node as ComponentNode).kind === 'component';
}

export async function resolveScreenRefs(params: {
  schema: ScreenSchema;
  loader: FragmentDocumentLoader;
  maxDepth?: number;
}): Promise<RefResolutionResult> {
  const { schema, loader, maxDepth = 32 } = params;

  const errors: RefResolutionError[] = [];
  const cache = new Map<string, ComponentNode>();

  const resolveNode = async (
    node: SchemaNode,
    ctx: { path: string; stack: string[]; depth: number },
  ): Promise<SchemaNode> => {
    if (ctx.depth > maxDepth) {
      errors.push({
        path: ctx.path,
        ref: ctx.stack.length === 0 ? '<depth>' : ctx.stack[ctx.stack.length - 1]!,
        message: `Exceeded maxDepth=${maxDepth}`,
      });
      return node;
    }

    if (isRefNode(node)) {
      const ref = node.ref;

      if (ctx.stack.includes(ref)) {
        errors.push({
          path: ctx.path,
          ref,
          message: `Circular reference: ${[...ctx.stack, ref].join(' -> ')}`,
        });
        return node;
      }

      const cached = cache.get(ref);
      if (cached) return cached;

      let raw: Record<string, unknown>;
      try {
        raw = await loader.loadFragmentDocument(ref);
      } catch (e) {
        errors.push({ path: ctx.path, ref, message: `Failed to load fragment: ${String(e)}` });
        return node;
      }

      const parsed = parseFragmentSchema(raw);
      if (!parsed.value) {
        errors.push({
          path: ctx.path,
          ref,
          message: `Invalid fragment document: ${parsed.errors.map((x) => `${x.path}: ${x.message}`).join('; ')}`,
        });
        return node;
      }

      const resolved = await resolveNode(parsed.value.node, {
        path: `${ctx.path}(ref:${ref})`,
        stack: [...ctx.stack, ref],
        depth: ctx.depth + 1,
      });

      if (isComponentNode(resolved)) {
        cache.set(ref, resolved);
      }

      return resolved;
    }

    if (isComponentNode(node)) {
      let didChange = false;
      const resolvedSlots: Record<string, SchemaNode[]> = {};

      for (const [slotName, children] of Object.entries(node.slots)) {
        const out: SchemaNode[] = [];
        for (let index = 0; index < children.length; index++) {
          const child = children[index]!;
          const resolvedChild = await resolveNode(child, {
            path: `${ctx.path}.slots.${slotName}[${index}]`,
            stack: ctx.stack,
            depth: ctx.depth,
          });
          out.push(resolvedChild);
          if (resolvedChild !== child) didChange = true;
        }
        resolvedSlots[slotName] = out;
      }

      if (!didChange) return node;

      return makeComponentNode({
        type: node.type,
        props: node.props,
        slots: resolvedSlots,
        actions: node.actions,
        bind: node.bind,
        visibleWhen: node.visibleWhen,
      });
    }

    return node;
  };

  const resolvedRoot = await resolveNode(schema.root, { path: 'root', stack: [], depth: 0 });

  const root = isComponentNode(resolvedRoot) ? resolvedRoot : schema.root;

  const nextSchema: ScreenSchema = {
    schemaVersion: schema.schemaVersion,
    id: schema.id,
    documentType: schema.documentType,
    product: schema.product,
    service: schema.service,
    themeId: schema.themeId,
    themeMode: schema.themeMode,
    root,
    actions: schema.actions,
  };

  return {
    schema: nextSchema,
    errors,
    isFullyResolved: errors.length === 0,
  };
}

File: packages/schema_runtime_dart/lib/src/schema/ref_resolver.dart
import 'schema_models.dart';
import 'schema_parser.dart';

abstract class FragmentDocumentLoader {
  Future<Map<String, Object?>> loadFragmentDocument(String fragmentId);
}

class RefResolutionError {
  const RefResolutionError({
    required this.path,
    required this.ref,
    required this.message,
  });

  final String path;
  final String ref;
  final String message;

  @override
  String toString() => '$path ($ref): $message';
}

class RefResolutionResult {
  const RefResolutionResult({required this.schema, required this.errors});

  final ScreenSchema schema;
  final List<RefResolutionError> errors;

  bool get isFullyResolved => errors.isEmpty;
}

Future<RefResolutionResult> resolveScreenRefs({
  required ScreenSchema schema,
  required FragmentDocumentLoader loader,
  int maxDepth = 32,
}) async {
  final errors = <RefResolutionError>[];
  final cache = <String, ComponentNode>{};

  Future<SchemaNode> resolveNode(
    SchemaNode node, {
    required String path,
    required List<String> stack,
    required int depth,
  }) async {
    if (depth > maxDepth) {
      errors.add(
        RefResolutionError(
          path: path,
          ref: stack.isEmpty ? '<depth>' : stack.last,
          message: 'Exceeded maxDepth=$maxDepth',
        ),
      );
      return node;
    }

    if (node is RefNode) {
      final ref = node.ref;
      if (stack.contains(ref)) {
        errors.add(
          RefResolutionError(
            path: path,
            ref: ref,
            message: 'Circular reference: ${[...stack, ref].join(' -> ')}',
          ),
        );
        return node;
      }

      final cached = cache[ref];
      if (cached != null) {
        return cached;
      }

      Map<String, Object?> raw;
      try {
        raw = await loader.loadFragmentDocument(ref);
      } catch (error) {
        errors.add(
          RefResolutionError(
            path: path,
            ref: ref,
            message: 'Failed to load fragment: $error',
          ),
        );
        return node;
      }

      final parsed = parseFragmentSchema(raw);
      final fragment = parsed.value;
      if (fragment == null) {
        errors.add(
          RefResolutionError(
            path: path,
            ref: ref,
            message: 'Invalid fragment document: ${parsed.errors.join('; ')}',
          ),
        );
        return node;
      }

      final resolved = await resolveNode(
        fragment.node,
        path: '$path(ref:$ref)',
        stack: [...stack, ref],
        depth: depth + 1,
      );

      if (resolved is ComponentNode) {
        cache[ref] = resolved;
      }
      return resolved;
    }

    if (node is ComponentNode) {
      var didChange = false;
      final resolvedSlots = <String, List<SchemaNode>>{};

      for (final entry in node.slots.entries) {
        final slotName = entry.key;
        final children = entry.value;
        final resolvedChildren = <SchemaNode>[];
        for (var index = 0; index < children.length; index++) {
          final child = children[index];
          final resolvedChild = await resolveNode(
            child,
            path: '$path.slots.$slotName[$index]',
            stack: stack,
            depth: depth,
          );
          resolvedChildren.add(resolvedChild);
          if (!identical(resolvedChild, child)) {
            didChange = true;
          }
        }
        resolvedSlots[slotName] = resolvedChildren;
      }

      if (!didChange) {
        return node;
      }

      return ComponentNode(
        type: node.type,
        props: node.props,
        slots: resolvedSlots,
        actions: node.actions,
        bind: node.bind,
        visibleWhen: node.visibleWhen,
      );
    }

    return node;
  }

  final resolvedRoot = await resolveNode(
    schema.root,
    path: 'root',
    stack: const [],
    depth: 0,
  );

  return RefResolutionResult(
    schema: ScreenSchema(
      schemaVersion: schema.schemaVersion,
      id: schema.id,
      documentType: schema.documentType,
      product: schema.product,
      service: schema.service,
      themeId: schema.themeId,
      themeMode: schema.themeMode,
      root: resolvedRoot is ComponentNode ? resolvedRoot : schema.root,
      actions: schema.actions,
    ),
    errors: errors,
  );
}
